primary_beam_correction.py

```diff
@@ -0,0 +1,296 @@
+    # Get positions of other sources
+    source_pos = wcs.WCS(images.header)
+    return phase_center, source_pos
+
+
+def radial_offset(phase_center, source_pos, row, col):
+    '''
+    This function gets the position in ra and dec of a pixel and determine its
+    radial offset from phase centre.
+    Input: phase center and source positions
+    output: A separation vector between phase centre and source positions in radians.
+    '''
+    p2w = source_pos.pixel_to_world(row, col, 0, 0)[0]
+    separation_rad = p2w.separation(phase_center).rad
+    return separation_rad
+
+
+def central_freq(path):
+    '''
+    Function to get central frequency of each frequency plane.
+    Input : image header
+    Output : List of central frequency in MHz of each frequency plane.
+    '''
+    images = read_fits(path)
+    c_freq_plane = []
+    for i in range(1, images.header['NSPEC']+1):
+        c_freq_plane.append(images.header['FREQ{0:04}'.format(i)])
+    return np.array(c_freq_plane)
+
+
+def beam_pattern(separation_rad, path, c_freq=None):
+    '''
+    Making beam pattern using the DEEP2 paper
+    Input : Radial seperation array and path to the fits file
+    Output : Beam pattern
+    '''
+    if c_freq:
+        rho = separation_rad
+        c = c_freq*1e6
+        # Degrees to radians conversion
+        conv = np.pi / 180.
+        # converting arcminutes to radians
+        v_beam_rad = 89.5 / 60. * conv
+        h_beam_rad = 86.2 / 60. * conv
+        # Taking the Geometric mean for the veritical and horizontal cut through the beam.
+        vh_beam_mean = np.sqrt(v_beam_rad*h_beam_rad)
+        flux_density = []
+        nu = c/1.e9
+        theta_b = vh_beam_mean*(1./nu)
+        ratio = rho/theta_b
+        num = np.cos(1.189*np.pi*ratio)
+        dem = 1-4*(1.189*ratio)**2
+        a_b = (num/dem)**2
+        flux_density.append(a_b)
+    else:
+        rho = separation_rad
+        c = central_freq(path)
+        # Degrees to radians conversion
+        conv = np.pi / 180.
+        # converting arcminutes to radians
+        v_beam_rad = 89.5 / 60. * conv
+        h_beam_rad = 86.2 / 60. * conv
+        # Taking the Geometric mean for the veritical and horizontal cut through the beam.
+        vh_beam_mean = np.sqrt(v_beam_rad*h_beam_rad)
+        flux_density = []
+        for nu in c:
+            nu = nu/1.e9
+            theta_b = vh_beam_mean*(1./nu)
+            ratio = rho/theta_b
+            num = np.cos(1.189*np.pi*ratio)
+            dem = 1-4*(1.189*ratio)**2
+            a_b = (num/dem)**2
+            flux_density.append(a_b)
+    return flux_density
+
+
+def mad(data):
+    '''
+    Calculating meadian absolute deviation (MAD).
+    '''
+    MAD_TO_SD = 1.4826
+    med = np.nanmedian(data)
+    dev = np.abs(data - med)
+    return med, MAD_TO_SD * np.nanmedian(dev)
+
+
+def standard_deviation(data):
+    '''
+    Calcalating the standard deviation of each frequency plane using MAD.
```

Comment:
-- and iterating. Rejecting pixels more than 5sigma from the mean until either no more pixels are rejected or a maximum of 50 iterations is reached.

Reply:
Done